kotlin/semlang-parser/src/main/kotlin/net/semlang/validator/validator.kt

```diff
@@ -736,41 +736,36 @@ private class Validator(
+        if (chosenParameter.isReference()) {
+            errors.add(Issue("Reference types cannot be used as parameters", expression.location, IssueLevel.ERROR))
         }
 
         val listType = Type.List(chosenParameter)
 
         val contents = expression.contents.map { item ->
-            validateExpression(item, variableTypes, typeParametersInScope, consumedThreadedVars, containingFunctionId) ?: return null
+            validateExpression(item, variableTypes, typeParametersInScope, refsAlreadyUsedInStatement, containingFunctionId) ?: return null
         }
         for (item in contents) {
             if (item.type != chosenParameter) {
                 error("Put an expression $item of type ${item.type} in a list literal of type ${listType}")
             }
         }
 
-        return TypedExpression.ListLiteral(listType, contents, chosenParameter)
+        return TypedExpression.ListLiteral(listType, AliasType.NotAliased, contents, chosenParameter)
     }
 
-    private fun validateIfThenExpression(expression: Expression.IfThen, variableTypes: Map<String, Type>, typeParametersInScope: Map<String, TypeParameter>, consumedThreadedVars: MutableSet<String>, containingFunctionId: EntityId): TypedExpression? {
-        val condition = validateExpression(expression.condition, variableTypes, typeParametersInScope, consumedThreadedVars, containingFunctionId) ?: return null
+    private fun validateIfThenExpression(expression: Expression.IfThen, variableTypes: Map<String, Type>, typeParametersInScope: Map<String, TypeParameter>, refsAlreadyUsedInStatement: MutableSet<String>, containingFunctionId: EntityId): TypedExpression? {
+        val condition = validateExpression(expression.condition, variableTypes, typeParametersInScope, refsAlreadyUsedInStatement, containingFunctionId) ?: return null
 
         if (condition.type != Type.BOOLEAN) {
             fail("In function $containingFunctionId, an if-then expression has a non-boolean condition expression: $condition")
         }
 
-        // Threaded variables can be consumed in each of the two blocks, but is considered used if used in either
-        val thenConsumedThreadedVars = HashSet(consumedThreadedVars)
-        val elseConsumedThreadedVars = HashSet(consumedThreadedVars)
-        val thenBlock = validateBlock(expression.thenBlock, variableTypes, typeParametersInScope, thenConsumedThreadedVars, containingFunctionId) ?: return null
-        val elseBlock = validateBlock(expression.elseBlock, variableTypes, typeParametersInScope, elseConsumedThreadedVars, containingFunctionId) ?: return null
-        // Don't add variable names that were defined in the blocks themselves
-        consumedThreadedVars.addAll(thenConsumedThreadedVars.intersect(variableTypes.keys))
-        consumedThreadedVars.addAll(elseConsumedThreadedVars.intersect(variableTypes.keys))
+        // TODO: Reconsider how references interact with if/then expressions
```

Comment:
This is notably a little weird right now in that if I make a ListBuilder in one block but put it in a variable, it becomes unaliasable for the if-statement as a whole, which is overly conservative. Solving this would probably involve tracking the actual set of aliases involved, not just aliased-or-not. This would be a good follow-up to fix.